### DISCUSSION TITLE
Entity property named 'data' does not work

### DISCUSSION THREAD

**Discussion by sebapatr96** [mikro-orm/mikro-orm#4406] [Q&A]
**Describe the bug**
If I declare an entity with field named "data" and tries to assign value to it upon creation, no value is being persisted there.

**To Reproduce**

This does not work:
```typescript
@Entity()
export class SomeEntity {
  @PrimaryKey()
  readonly id: number;

  @Property({ type: JsonType, columnType: 'jsonb', nullable: true })
  data?: Record<string, any>;
}
```

This works:
```typescript
@Entity()
export class SomeEntity {
  @PrimaryKey()
  readonly id: number;

  @Property({ type: JsonType, columnType: 'jsonb', fieldName: 'data', nullable: true })
  random?: Record<string, any>;
}
```

**Versions**

| Dependency | Version |
| - | - |
| node | 18.15.0 |
| typescript | 4.9.3 |
| mikro-orm |5.7.10 |

**Comment by B4nan** [comment]
Can you provide actual reproduction? Entity definition is not saying much, provide the code of what you are doing with the entity.

I can assure you, there are no restrictions on property names, your problem will be rather in out-of-sync metadata/cache/schema. There are even some existing tests using `data` as a property name (of JSON type).

**Reply by sebapatr96** [reply]
Ok I looked into this issue more closely now, sorry for misleading in the first post, you were right, the issue is somewhere else. The issue seems to be when mikro-orm calls internaly entity's constructor, when calling .create method on entityManager's instance like

```typescript
em.create(SomeEntity, {
    data: {
      random: 123
    },
});
```

in this case, Mikro-ORM calls constructor of SomeEntity with the data (I mean the whole data object from second argument of .create, not just the data property of it) passed on the first argument of constructor. So if I have entity's constructor like this

```typescript
@Entity()
export class SomeEntity {
  constructor(dto: {
    data?: Record<string, any>;
  }) {
    this.data = data;
  }
  
  @PrimaryKey()
  readonly id: number;

  @Property({ type: JsonType, columnType: 'jsonb', nullable: true })
  data?: Record<string, any>;
}
```

the `dto.data` is left undefined. Is this intentional? I was looking into docs right [here](https://mikro-orm.io/docs/entity-constructors#pojo-vs-entity-instance-in-constructor) but did't find anything.

I know there should not be reason to just call `new` right away, which of course would avoid this issue as I now know where its coming from. But I'd rather inform you where the issue came from
```typescript
new SomeEntity({
  data: {
    random: 123
  }
})
```

**Reply by B4nan** [reply]
> in this case, Mikro-ORM calls constructor of SomeEntity with the data (I mean the whole data object from second argument of .create, not just the data property of it) passed on the first argument of constructor. So if I have entity's constructor like this

If your constructor parameters match the property names, you will get the property values on corresponding positions, otherwise, the first argument will be the full data object. That's how it is supposed to work, so if your property is `data` and your constructor parameter is `dto`, you should get the full params object in the first position - if you would call the parameter just `data`, you should get only the dictionary.

This is the place where this happens: 

https://github.com/mikro-orm/mikro-orm/blob/master/packages/core/src/entity/EntityFactory.ts#L205-L207
